data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java

```diff
@@ -3,32 +3,30 @@
 
 package com.microsoft.azure.kusto.data;
 
+import com.azure.core.http.HttpClient;
+import com.azure.core.http.HttpHeaderName;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.node.ObjectNode;
 import com.microsoft.azure.kusto.data.auth.CloudInfo;
 import com.microsoft.azure.kusto.data.auth.ConnectionStringBuilder;
 import com.microsoft.azure.kusto.data.auth.TokenProviderBase;
 import com.microsoft.azure.kusto.data.auth.TokenProviderFactory;
 import com.microsoft.azure.kusto.data.auth.endpoints.KustoTrustedEndpoints;
-import com.microsoft.azure.kusto.data.exceptions.DataClientException;
-import com.microsoft.azure.kusto.data.exceptions.DataServiceException;
-import com.microsoft.azure.kusto.data.exceptions.KustoClientInvalidConnectionStringException;
-import com.microsoft.azure.kusto.data.exceptions.KustoServiceQueryError;
+import com.microsoft.azure.kusto.data.exceptions.*;
 import com.microsoft.azure.kusto.data.http.HttpClientFactory;
+import com.microsoft.azure.kusto.data.http.HttpClientProperties;
 import com.microsoft.azure.kusto.data.http.UncloseableStream;
 import com.microsoft.azure.kusto.data.http.HttpPostUtils;
 import com.microsoft.azure.kusto.data.instrumentation.MonitoredActivity;
 import com.microsoft.azure.kusto.data.instrumentation.SupplierTwoExceptions;
 import com.microsoft.azure.kusto.data.instrumentation.TraceableAttributes;
 import org.apache.commons.lang3.StringUtils;
-import org.apache.http.HttpHeaders;
-import org.apache.http.ParseException;
-import org.apache.http.client.utils.URIBuilder;
-import org.apache.http.entity.AbstractHttpEntity;
-import org.apache.http.entity.ContentType;
-import org.apache.http.entity.InputStreamEntity;
-import org.apache.http.entity.StringEntity;
-import org.apache.http.impl.client.CloseableHttpClient;
+
+import org.apache.hc.core5.http.ContentType;
```

Comment:
Why do we need to have this dependency? Isn't the whole point is to get rid of the apache http server?

I don't think (from what I've seen in the code) we need to use the "Entity" abstraction - the BinaryData abstraction in the Azure package should be good enough.

And URIBuilder, we can't achieve this otherwise?

Reply:
@AsafMah I removed references to entity and replaced it with BinaryData from the Azure Core library. The URIBuilder is a little tricky. I will need to dig into how it is implemented before I can remove it, but I will do that next.

I know this is supposed to be a minimal implementation, but there's another issue related to HTTP request generation that I'd like to ask for permission to explore.

I'd really like to see the post methods exist on the KustoClient itself, that way the post methods could just use the KustoClient's inner HTTP client without the HTTP client needing passed around to the methods. This would help shorten method signatures.

I'd also really like to create an HttpRequestBuilder class and move the request building logic out of HttpPostUtils and out of the client implementations and into a request builder class. 

This would greatly shorten the post method signatures on HttpPostUtils and move request building logic out of the clients and into a class dedicated for that purpose, providing more logical separation of that phase.

e.g.

Phase 1: Parse/Validate Input and Generate an HTTP Request (builder class, separately testable)
Phase 2: Send request (monitored activity using KustoClient's inner HttpClient)
Phase 3: Parse/Validate Response and convert to KustoResult (potentially a separate builder?)

In this setup the post methods take a ready-made Request object, the Request object is then sent via the overarching client abstraction's inner HTTP client, and then the response is parsed.

With your and your team's permission I'd really like to explore that possibility in this PR. I think it would really help readability for newcomers to the codebase. I would of course adjust tests to account for the change.

Reply:
I started working on the above just to see what it would do with the codebase and it looks really really nice and has a lot of positive impacts in the client class. I haven't committed any of it. If you wanted I could open a separate branch, but I feel **much** more confident about following HTTP request building logic with everything in one place. I'd really like to add this.